internal/kms/kmip.go

```diff
@@ -0,0 +1,512 @@
+	caCertPool.AppendCertsFromPEM([]byte(caCert))
+	cert, err := tls.X509KeyPair([]byte(clientCert), []byte(clientKey))
+	if err != nil {
+		return nil, fmt.Errorf("invalid X509 key pair: %w", err)
+	}
+
+	kms.tlsConfig = &tls.Config{
+		MinVersion:   tls.VersionTLS12,
+		ServerName:   serverName,
+		RootCAs:      caCertPool,
+		Certificates: []tls.Certificate{cert},
+	}
+
+	return kms, nil
+}
+
+// EncryptDEK uses the Amazon KMS and the configured CMK to encrypt the DEK.
+func (kms *kmipKMS) EncryptDEK(_, plainDEK string) (string, error) {
+	conn, err := kms.connect()
+	if err != nil {
+		return "", err
+	}
+	defer conn.Close()
+
+	emd := encryptedMetedataDEK{}
+	emd.Nonce, err = generateNonce(nonceSize)
+	if err != nil {
+		return "", fmt.Errorf("failed to generated nonce: %w", err)
+	}
+
+	respMsg, decoder, uniqueBatchItemID, err := kms.send(conn,
+		kmip14.OperationEncrypt,
+		EncryptRequestPayload{
+			UniqueIdentifier: kms.uniqueIdentifier,
+			Data:             []byte(plainDEK),
+			CryptographicParameters: kmip.CryptographicParameters{
+				PaddingMethod:          kmip14.PaddingMethodPKCS5,
+				CryptographicAlgorithm: kmip14.CryptographicAlgorithmAES,
+				BlockCipherMode:        kmip14.BlockCipherModeCBC,
+			},
+			IVCounterNonce: emd.Nonce,
+		})
+	if err != nil {
+		return "", err
+	}
+
+	bi, err := kms.response(respMsg, kmip14.OperationEncrypt, uniqueBatchItemID)
+	if err != nil {
+		return "", err
+	}
+
+	ttlvPayload, ok := bi.ResponsePayload.(ttlv.TTLV)
+	if !ok {
+		return "", errors.New("failed to parse responsePayload")
+	}
+
+	var encryptRespPayload EncryptResponsePayload
+	err = decoder.DecodeValue(&encryptRespPayload, ttlvPayload)
+	if err != nil {
+		return "", err
+	}
+
+	emd.DEK = encryptRespPayload.Data
+	emdData, err := json.Marshal(&emd)
+	if err != nil {
+		return "", fmt.Errorf("failed to convert "+
+			"encryptedMetedataDEK to JSON: %w", err)
+	}
+
+	return string(emdData), nil
+}
+
+// DecryptDEK uses the Amazon KMS and the configured CMK to decrypt the DEK.
+func (kms *kmipKMS) DecryptDEK(_, encryptedDEK string) (string, error) {
+	conn, err := kms.connect()
+	if err != nil {
+		return "", err
+	}
+	defer conn.Close()
+
+	emd := encryptedMetedataDEK{}
+	err = json.Unmarshal([]byte(encryptedDEK), &emd)
+	if err != nil {
+		return "", fmt.Errorf("failed to convert data to "+
+			"encryptedMetedataDEK: %w", err)
+	}
+
+	respMsg, decoder, uniqueBatchItemID, err := kms.send(conn,
+		kmip14.OperationDecrypt,
+		DecryptRequestPayload{
+			UniqueIdentifier: kms.uniqueIdentifier,
+			Data:             emd.DEK,
+			IVCounterNonce:   emd.DEK,
+			CryptographicParameters: kmip.CryptographicParameters{
+				PaddingMethod:          kmip14.PaddingMethodPKCS5,
+				CryptographicAlgorithm: kmip14.CryptographicAlgorithmAES,
+				BlockCipherMode:        kmip14.BlockCipherModeCBC,
+			},
+		})
+	if err != nil {
+		return "", err
+	}
+
+	bi, err := kms.response(respMsg, kmip14.OperationDecrypt, uniqueBatchItemID)
```

Comment:
what is `bi`? use a better name with more readability?

Reply:
renaming to batchItem